src/HallPIC.jl

```diff
@@ -630,22 +630,47 @@ struct Reaction{T<:AbstractFloat}
     threshold_energy::T
     rate_table::LinearInterpolation
     delta_n::Vector{T}
+    delta_n_remainder::Vector{T}
 
 end
 
 
 
-function deplete_reactant!(reaction::Reaction{T}, reactant::ParticleContainer{T}, fluid_properties::SpeciesProperties{T}, grid::Grid, electron_properties::SpeciesProperties{T}, dt) where T
+function deplete_reactant!(reaction::Reaction{T}, reactant::ParticleContainer{T}, reactant_properties::SpeciesProperties{T}, products::Vector{ParticleContainer{T}}, product_properties::Vector{SpeciesProperties{T}}, grid::Grid, electron_properties::SpeciesProperties{T}, dt) where T
 
     # for each (non-ghost) cell, calculate the delta n
     for i in 2:length(grid.cell_centers)-1
 
         # find the rate from a lookup table 
         rate = reaction.rate_table(electron_properties.temp[i])
         # calculate the number of particles produced
-        reaction.delta_n[i] = fluid_properties.dens[i] * electron_properties.dens[i] * rate * dt
+        reaction.delta_n[i] = (reactant_properties.dens[i] * electron_properties.dens[i] * rate * dt) + reaction.delta_n_remainder[i]
     end
 
+    # loop over the products to compute the weight consumed 
+    # for each cell, add particles 
+    for i in 2:length(grid.cell_centers)-1
+        n_consumed = Inf
+        for (ip, product) in enumerate(product_properties)
+
+            n_desired = products[ip].n_d # number of desired particles touching cell, hyperparameter from the simulation
+        
+            # determine the number of partices to generate
+            w_gen = product.avg_weight[i] * (product.N_particles[i]/n_desired) # check for particles in the cell 
+            real_particles_produced = reaction.product_coefficients[ip] * reaction.delta_n[i] 
+            n_gen = Int32(floor(real_particles_produced / (w_gen)))
+
+            n_consumed = minimum([n_consumed, n_gen * w_gen])            
+
+        end
+
+        #set the remainder and the delta n 
+        reaction.delta_n_remainder[i] = reaction.delta_n[i] - n_consumed 
+        reaction.delta_n[i] = n_consumed
+
+    end
```

Comment:
can you invert this loop? i.e. loop over product species then over cells? should be more cache-friendly